src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1505,12 +1636,179 @@ namespace Microsoft.FSharp.Core
+                        member __.Equals (x,y) = (# "ceq" x y : bool #) || (not ((# "ceq" x x : bool #) || (# "ceq" y y : bool #)))
+                        member __.GetHashCode x = x.GetHashCode () }
+
+                | true, ty when ty.Equals typeof<float32> -> box {
+                    new EqualityComparer<float32>() with
+                        member __.Equals (x,y) = (# "ceq" x y : bool #) || (not ((# "ceq" x x : bool #) || (# "ceq" y y : bool #)))
+                        member __.GetHashCode x = x.GetHashCode () }
+
+                | _ -> null
+
+            let genericFSharpEqualityComparer_ER<'T> () = {
+                new EqualityComparer<'T>() with
+                    member __.Equals (x,y) = GenericEqualityObj true fsEqualityComparerUnlimitedHashingER (box x, box y)
+                    member __.GetHashCode x = GenericHashParamObj fsEqualityComparerUnlimitedHashingPER (box x)
+            }
+
+            let genericFSharpEqualityComparer_PER<'T> () = {
+                new EqualityComparer<'T>() with
+                    member __.Equals (x,y) = GenericEqualityObj false fsEqualityComparerUnlimitedHashingPER (box x, box y)
+                    member __.GetHashCode x = GenericHashParamObj fsEqualityComparerUnlimitedHashingPER (box x)
+            }
+
+            let getGenericEquality<'T> er =
+                match tryGetFSharpEqualityComparer er typeof<'T> with
+                | :? EqualityComparer<'T> as call -> call
+                | _ when canUseDefaultEqualityComparer typeof<'T> -> EqualityComparer<'T>.Default
+                | _ when er -> genericFSharpEqualityComparer_ER<'T> ()
+                | _ -> genericFSharpEqualityComparer_PER<'T> ()
+
+            type FSharpEqualityComparer_ER<'T> private () =
+                static let comparer = getGenericEquality<'T> true
+
+                static member Comparer = comparer
+
+                static member inline Equals (x, y) = comparer.Equals (x, y)
+                static member inline GetHashCode x = comparer.GetHashCode x
+
+            type FSharpEqualityComparer_PER<'T> private () =
+                static let comparer = getGenericEquality<'T> false
+
+                static member Comparer = comparer
+
+                static member inline Equals (x, y) = comparer.Equals (x, y)
+                static member inline GetHashCode x = comparer.GetHashCode x
+
             /// Implements generic equality between two values, with PER semantics for NaN (so equality on two NaN values returns false)
             //
             // The compiler optimizer is aware of this function  (see use of generic_equality_per_inner_vref in opt.fs)
             // and devirtualizes calls to it based on "T".
             let GenericEqualityIntrinsic (x : 'T) (y : 'T) : bool = 
-                GenericEqualityObj false fsEqualityComparerNoHashingPER ((box x), (box y))
+                FSharpEqualityComparer_PER<'T>.Equals (x, y)
```

Comment:
I'm aware that this is now more expensive initially: we have the table lookup followed by an indirect call, before getting to GenericEqualityObj.  

Am I correct that that's the fundamental tradeoff inherent to the PR?  i.e. those costs are small compared to subsequent savings of not boxing (for primitive value types) and the advantages of using the default equality comparer (for other types)?

Reply:
There is a slight cost, but the cost of boxing (for value types) and type-based pattern matching (in GenericEqualityObj) are an order of magnitude higher  (he says without actually running the numbers, but from general feel.) 

Also remember that these functions are only kicking in when equality doesn't know the type at compile time, so it's not a hit all over the place.